### ISSUE TITLE
Setting yarn cache folder with relative path does not resolve to absolute path

### ISSUE THREAD

**Issue by raymondwang** [yarnpkg/yarn#2881]
**Do you want to request a *feature* or report a *bug*?**
I think this is a bug. If this is actually intended behavior (which I can maybe see as being useful when switching between multiple projects and wanting a different yarn cache directory for each at some consistent relative path), happy to close.

**What is the current behavior?**
If you set the yarn cache directory using a relative path, yarn never resolves an absolute reference for that directory. For example, if you run `yarn config set cache-folder ../yarn-cache`, `yarn cache dir` will output `../yarn-cache`, no matter where you are.

**If the current behavior is a bug, please provide the steps to reproduce.**
Run `yarn config set cache-folder ../yarn-cache`. `../yarn-cache` will be created, but `yarn cache dir` will still output `../yarn-cache`, and if you change directories then run `yarn` again, a new folder will be made at `../yarn-cache`.

**What is the expected behavior?**
The cache folder should resolve to the absolute path for the given directory, i.e. if I set the cache folder to `../yarn-cache` from `~/foo/bar/`, I would expect the output of `yarn cache dir` to become `~/foo/yarn-cache`.

**Please mention your node.js, yarn and operating system version.**
Node v7.7.2, yarn 0.21.3, Mac OS 10.12.3

**Comment by raymondwang** [comment]
Please ignore the title changes. I faked myself out by thinking I'd misreported the problem when I noticed something else was buggy, but then realized I'd just gotten myself into a weird state in my local environment while inadvertently becoming a yarn cache directory factory.

**Comment by clanghout** [comment]
I think the bug consists of 2 parts (I ran into both working on #2870): 
- In my pr, the relative path to the yarn bin folder pointed to the wrong depth of the folder structure happening only on Mac OS.
- Changing the cachefolder is currently not correctly changing the config.

For the first problem I switched to using the absolute path instead. For the second problem the fix is in the pr and can be found [on this line](https://github.com/yarnpkg/yarn/pull/2870/files#diff-867becf4a9c2c6c6d4e7c1278750724eR365).
